### ISSUE TITLE
Streaming HttpContent and ReadAsStreamAsync

### ISSUE THREAD

**Issue by JamesNK** [dotnet/corefx#42160]
The default implementation of `HttpContent.ReadAsStreamAsync` copies content into a buffer stream and doesn't return until `HttpContent.SerializeToStreamAsync` is complete. This becomes a problem with streaming implementations, as someone could called `ReadAsStreamAsync` expecting to get content as it is written rather than waiting until the stream is complete.

Related issue in how ASP.NET Core's TestServer using HttpContent: https://github.com/aspnet/AspNetCore/pull/15591

I'm searching for guidance and discussion here about this situation.
1. Is the current default buffer behavior correct? - https://github.com/dotnet/corefx/blob/26d7b6626ab0385a1158e674855f81ab73f628e0/src/System.Net.Http/src/System/Net/Http/HttpContent.cs#L453-L459
Maybe it should do something different when AllowDuplex is set to true - https://github.com/dotnet/corefx/blob/26d7b6626ab0385a1158e674855f81ab73f628e0/src/System.Net.Http/src/System/Net/Http/HttpContent.cs#L320-L326
Or perhaps there should be a new abstract HttpContent that is designed for duplex.
2. Or should we rely on all [custom streaming implementations](https://github.com/grpc/grpc-dotnet/blob/9e39edb0a3fc7a625085490bf64c2794870a519e/src/Grpc.Net.Client/Internal/PushStreamContent.cs) overriding `HttpContent.CreateContentReadStreamAsync`? It's relatively simple to do using a `Pipe` and Stream adapters with the reader and writer (see linked PR in ASP.NET Core), but it forces the implementation to take a dependency on all of Pipes just in case someone calls `ReadAsStreamAsync` in a delegating handler.
3. Not worry about it.

**Comment by scalablecory** [comment]
I would be interested in a `ReadAsStreamAsync` that explicitly doesn't buffer.

Related https://github.com/dotnet/corefx/issues/38559

**Comment by davidsh** [comment]
>I would be interested in a ReadAsStreamAsync that explicitly doesn't buffer.

From a design perspective HttpContent.ReadAsStreamAsync() never buffers. It simply returns the stream.

When we  originally designed the HttpContent APIs for ReadAs*, we made them all have the "async" suffix and return a Task. That was required for ReadAsStringAsync, ReadAsByteArrayAsync since they are explicitly reading (buffering) in the entire response body. But ReadAsStreamAsync was never going to buffer, ever.

In fact, we debated even having it be an async Task API. It doesn't have to be really. It could been called "ReadAsStream" or even "GetStream".  But we settled on consistency in the naming and API pattern.

We'd had feedback that perhaps we should add an HttpContent.ReadAsStream or equivalent API because that is really what it does and it doesn't have to be async (there is no async I/O required to return a stream back).

**Comment by JamesNK** [comment]
> From a design perspective HttpContent.ReadAsStreamAsync() never buffers. It simply returns the stream.

Yes it does? That's what I've observed, and the source code used to return the stream loads all the content from `SerializeToStreamAsync` into a stream, rewinds it, then returns that.

https://github.com/dotnet/corefx/blob/26d7b6626ab0385a1158e674855f81ab73f628e0/src/System.Net.Http/src/System/Net/Http/HttpContent.cs#L453-L459

`StreamContent.ReadAsStreamAsync()` does just return its stream, but that's just one implementation.

**Comment by davidsh** [comment]
>Yes it does? That's what I've observed, and the source code used to return the stream loads all the content from SerializeToStreamAsync into a stream, rewinds it, then returns that.

Sorry. In thinking about this now, perhaps I mixed these up.  I should not have talked about HttpContent.ReadAsStreamAsync().

Instead, I think I was really referring to HttpClient.GetStreamAsync() where that API never buffers the response body. And that was the one that could have been a pure sync API (HttpClient.GetStream) since there isn't any I/O activity to do that.  But in designing HttpClient convenience methods, we made all of the async even though GetStreamAsync() didn't have to be async.

**Comment by stephentoub** [comment]
@JamesNK, what is the scenario where this is showing up for you as an issue?

Request content is meant to be sent on the wire using HttpContent.CopyToAsync, e.g. that's how SocketsHttpHandler does it:
https://github.com/dotnet/corefx/blob/6e1d3ff966d72b2e9903ce0b208403eaa9e492ff/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs#L781-L782
and that just uses SerializeToStreamAsync, no full-stream buffering:
https://github.com/dotnet/corefx/blob/6e1d3ff966d72b2e9903ce0b208403eaa9e492ff/src/System.Net.Http/src/System/Net/Http/HttpContent.cs#L349

Response content is implemented by the handler.  A good implementation will override CreateContentReadStreamAsync, e.g. https://github.com/dotnet/corefx/blob/6e1d3ff966d72b2e9903ce0b208403eaa9e492ff/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionResponseContent.cs#L57-L58, as otherwise it'll rely on the base implementation, which as you highlighted will use SerializeToStreamAsync, and thus will serialize it into a MemoryStream that can then be returned (if it tried to do some kind of producer/consumer stream instead, it would be a form of fire-and-forget).

Everything appears to be behaving as it should.  Implementations of HttpClient override the abstract SerializeToStreamAsync, and get a "free" CreateContentReadStreamAsync... if they want a better one, they override it, and most of our built-in implementations do.

What is the concrete problem?

**Comment by JamesNK** [comment]
The concrete problem is the `TestServer` in ASP.NET Core (in-memory simulation of a server) was reading `ReadAsStreamAsync`. I have a PR in progress to replace `ReadAsStreamAsync` with a pipe and populate it using `HttpContent.CopyToAsync`.

https://github.com/aspnet/AspNetCore/pull/15591

This fixes `TestServer`. I created this issue to check whether the current buffer behavior is still correct now that duplex streaming is a thing with HttpClient+HTTP/2.

**Comment by stephentoub** [comment]
> was reading ReadAsStreamAsync

On what kind of HttpContent?

**Comment by JamesNK** [comment]
https://github.com/grpc/grpc-dotnet/blob/9e39edb0a3fc7a625085490bf64c2794870a519e/src/Grpc.Net.Client/Internal/PushStreamContent.cs

**Comment by stephentoub** [comment]
Thanks.  Yeah, your PushHttpContent is explicitly a producer in a producer/consumer scenario.  You would need to override CreateContentReadStreamAsync to similarly implement a consumer.

**Comment by JamesNK** [comment]
So option 2 from https://github.com/dotnet/corefx/issues/42160#issue-513014040

It's relatively simple to do using a `Pipe` and Stream adapters with the reader and writer, but it forces the implementation to take a dependency on all of Pipes just in case someone calls ReadAsStreamAsync in a delegating handler.

Is there a better way? Is there anything that could be done in corefx to not put the complexity of implementing pub/sub streams on developers.

**Comment by JamesNK** [comment]
Or perhaps CreateContentReadStreamAsync should just throw an exception if AllowDuplex is true. A clear exception message saying "ReadAsStreamAsync is not supported, you should call CopyToAsync" would be preferable to accidentally deadlocking by buffering.
